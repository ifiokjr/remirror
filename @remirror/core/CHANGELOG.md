# @remirror/core

## 0.8.0

### Minor Changes

- 24f83413: - Change the signature of `CommandFunction` to only accept one parameter which contains
  `dispatch`, `view`, `state`.

  - Create a new exported `ProsemirrorCommandFunction` type to describe the prosemirror commands which are
    still used in the codebase.
  - Rename `KeyboardBindings` to `KeyBindings`. Allow `CommandFunctionParams` to provide extra parameters like
    `next` in the newly named `KeyBindings`.
  - Create a new `KeyBindingCommandFunction` to describe the `Extension.keys()` return type. Update the name
    of the `ExcludeOptions.keymaps` -> `ExcludeOptions.keys`.

  **BREAKING CHANGE**

- 24f83413: Improve the way `ExtensionManager` calls `Extension.keys` methods. Keys now use the new api for
  CommandFunctions which provides a `next` method. This method allows for better control when deciding whether
  or not to defer to the next keybinding in the chain.

  To override, create a new keybinding with another extension. Make sure the extension is created with a
  higher priority. The keybinding you create can either return true or false. By default if it returns true,
  no other keybindings will run. However if it returns `false` all other keybindings will be run until one
  returns `true`

  `next` basically calls the every lower priority keybinding in the extensions list. It gives you full control
  of how the bindings are called.

### Patch Changes

- Updated dependencies [24f83413]
  - @remirror/core-types@0.8.0
  - @remirror/core-helpers@0.7.5
  - @remirror/core-utils@0.7.5
  - prosemirror-suggest@0.7.5

## 0.7.4

### Patch Changes

- 7380e18f: Update repository url from ifiokjr/remirror to remirror/remirror to reflect new GitHub
  organisation.
- Updated dependencies [7380e18f]
  - @remirror/core-constants@0.7.4
  - @remirror/core-helpers@0.7.4
  - @remirror/core-types@0.7.4
  - @remirror/core-utils@0.7.4
  - @remirror/react-portals@0.7.4
  - prosemirror-suggest@0.7.4

## 0.7.3

### Patch Changes

- 5f85c0de: Bump a new version to test out the changeset API.
- Updated dependencies [5f85c0de]
  - @remirror/core-helpers@0.7.3
  - @remirror/core-constants@0.7.3
  - @remirror/core-types@0.7.3
  - @remirror/core-utils@0.7.3
  - @remirror/react-portals@0.7.3
  - prosemirror-suggest@0.7.3
