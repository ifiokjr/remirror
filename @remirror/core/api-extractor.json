{
  "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
  "mainEntryPointFile": "lib/index.d.ts",
  "apiReport": {
    "enabled": true,
    "reportFolder": "etc"
  },
  "docModel": {
    "enabled": true
  },
  "dtsRollup": {
    "enabled": true,
    "untrimmedFilePath": "lib/dist/core-alpha.d.ts",
    "betaTrimmedFilePath": "lib/dist/core-beta.d.ts",
    "publicTrimmedFilePath": "lib/dist/core.d.ts"
  }
}
