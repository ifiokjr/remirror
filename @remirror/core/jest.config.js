const config = require('../../config/jest.config');

module.exports = {
  ...config,
  name: '@remirror/core',
  displayName: 'core',
};
