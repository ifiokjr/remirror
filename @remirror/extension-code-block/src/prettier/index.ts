// import { Plugin, formatWithCursor } from 'prettier/standalone'
// import { SchemaNodeTypeParams } from '@remirror/core';

// /**
//  * Prettier is a very large bundle size. 1.2mb GZipped all inclusive. So it's up
//  * to you whether you want to include this or not.
//  *
//  * It allows for formatting markdown, typescript and javascript code blocks
//  *
//  * ```ts
//  *
//  * ```
//  *
//  */
// export function markdownCommand(plugins: Plugin[]) {
//   return (params: SchemaNodeTypeParams) => {
//     return {

//     }
//   }
// }
