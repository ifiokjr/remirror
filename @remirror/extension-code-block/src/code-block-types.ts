import { Attrs, NodeExtensionOptions } from '@remirror/core';
import { RefractorSyntax } from 'refractor/core';

import { SyntaxTheme } from './themes';

export interface CodeBlockExtensionOptions extends NodeExtensionOptions {
  /**
   * Import languages from refractor
   *
   * @remarks
   *
   * ```ts
   * import jsx from 'refractor/lang/jsx'
   * import typescript from 'refractor/lang/typescript'
   * ```
   *
   * And pass them into the config when initializing this extension.
   *
   * ```ts
   * import { CodeBlockExtension } from '@remirror/extension-code-block';
   *
   * new CodeBlockExtension({ supportedLanguages: [typescript, jsx] })
   * ```
   *
   * Or as a component
   *
   * ```tsx
   * <RemirrorManager>
   *   <RemirrorExtension Constructor={CodeBlockExtension} supportedLanguages={[typescript, jsx]} />
   * </RemirrorManager>
   * ```
   *
   * By default refractor bundles the following languages: `markup`, `css`, `clike`, `js`
   *
   * @defaultValue `[]`
   */
  supportedLanguages?: RefractorSyntax[];

  /**
   * The default language to use when none is provided.
   *
   * @defaultValue 'markup'
   */
  defaultLanguage?: string;

  /**
   * The theme to use for the codeBlocks.
   *
   * @remarks
   * Currently only one theme can be set per editor.
   *
   * Set this to false if you want to manage the syntax styles by yourself.
   * For tips on how this could be accomplished see {@link https://prismjs.com}
   *
   * @defaultValue 'atomDark'
   */
  syntaxTheme?: SyntaxTheme | false;

  /**
   * Provide a formatter which can format the provided source code.
   *
   * @returns an object when formatting was successful and false when the code could not be formatted (a noop).
   */
  formatter?: CodeBlockFormatter;

  /**
   * A keyboard shortcut to trigger formatting the current block.
   *
   * @defaultValue `Alt-Shift-F` (Mac) `Shift-Ctrl-F` (PC)
   */
  keyboardShortcut?: string;

  /**
   * The name of the node that the code block should toggle back and forth from.
   *
   * @defaultValue 'paragraph'
   */
  toggleType?: string;
}

/**
 * A function which takes code and formats the code.
 *
 * TODO - possibly allow error management if failure is because of invalid syntax
 */
export type CodeBlockFormatter = (params: FormatterParams) => FormattedContent | undefined;

export interface FormatterParams {
  /**
   * The code to be formatted
   */
  source: string;

  /**
   * Specify where the cursor is. This option cannot be used with rangeStart and rangeEnd.
   * This allows the command to both formats the code, and translates a cursor position from unformatted code to formatted code.
   */
  cursorOffset: number;

  /**
   * The language of the code block. Should be used to determine whether the formatter can support the transformation.
   *
   * Possible languages are available here https://github.com/wooorm/refractor/tree/716fe904c37cd7ebfde53ac5157e7d6c323a3986/lang
   */
  language: string;
}

/**
 * Data returned from a code formatter.
 */
export interface FormattedContent {
  /**
   * The transformed source.
   */
  formatted: string;

  /**
   * The new cursor position after formatting
   */
  cursorOffset: number;
}

export interface CodeBlockAttrs extends Attrs {
  /**
   * The language attribute
   */
  language: string;
}
