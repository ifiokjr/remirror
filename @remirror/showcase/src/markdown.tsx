/** @jsx jsx */

import { jsx } from '@emotion/core';
import { capitalize } from '@remirror/core';
import { EditorDisplay, MarkdownEditor } from '@remirror/editor-markdown';
import { Button } from '@remirror/ui-buttons';
import { FC, useState } from 'react';

const initialValue = {
  type: 'doc',
  content: [
    { type: 'heading', attrs: { level: 1 }, content: [{ type: 'text', text: 'Markdown Editor' }] },
    {
      type: 'paragraph',
      content: [
        { type: 'text', text: 'With ' },
        { type: 'text', marks: [{ type: 'bold' }, { type: 'italic' }], text: 'formatting ' },
        { type: 'text', text: 'and other ' },
        { type: 'text', marks: [{ type: 'italic' }], text: 'enhancements' },
        { type: 'text', text: '.' },
      ],
    },
    {
      type: 'codeBlock',
      attrs: { language: 'js' },
      content: [
        {
          type: 'text',
          text: 'const a = "welcome"',
        },
      ],
    },
  ],
};

export const ExampleMarkdownEditor: FC = ({ children }) => {
  const [editor, setEditor] = useState<EditorDisplay>('markdown');

  const toggleEditor = () => {
    if (editor === 'wysiwyg') {
      setEditor('markdown');
    }
    if (editor === 'markdown') {
      setEditor('wysiwyg');
    }
  };

  return (
    <div>
      <span>
        <Button onClick={toggleEditor} content='Toggle Editor' variant='primary' /> {capitalize(editor)}
      </span>
      <MarkdownEditor editor={editor} initialValue={initialValue}>
        {children}
      </MarkdownEditor>
    </div>
  );
};
