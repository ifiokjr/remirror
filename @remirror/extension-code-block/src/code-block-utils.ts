import {
  Attrs,
  bool,
  EditorState,
  findParentNodeOfType,
  flattenArray,
  FromToParams,
  isEqual,
  isObject,
  isString,
  NodeType,
  NodeTypeParams,
  NodeWithPosition,
  PosParams,
  ProsemirrorCommandFunction,
  ProsemirrorNodeParams,
  TextParams,
  uniqueArray,
} from '@remirror/core';
// Refractor languages
import { TextSelection } from 'prosemirror-state';
import { Decoration } from 'prosemirror-view';
import refractor, { RefractorNode, RefractorSyntax } from 'refractor/core';
import clike from 'refractor/lang/clike';
import css from 'refractor/lang/css';
import js from 'refractor/lang/javascript';
import markup from 'refractor/lang/markup';

import { CodeBlockAttrs, CodeBlockExtensionOptions, FormattedContent } from './code-block-types';

interface ParsedRefractorNode extends TextParams {
  /**
   * The classes that will wrap the node
   */
  classes: string[];
}

interface PositionedRefractorNode extends FromToParams, ParsedRefractorNode {}

/**
 * Maps the refractor nodes into text and classes which will be used to create our decoration.
 */
function parseRefractorNodes(
  refractorNodes: RefractorNode[],
  className: string[] = [],
): ParsedRefractorNode[][] {
  return refractorNodes.map(node => {
    const classes = [
      ...className,
      ...(node.type === 'element' && node.properties.className ? node.properties.className : []),
    ];

    if (node.type === 'element') {
      return parseRefractorNodes(node.children, classes) as any;
    }

    return {
      text: node.value,
      classes,
    };
  });
}

interface CreateDecorationsParams extends Pick<CodeBlockExtensionOptions, 'defaultLanguage'> {
  /**
   * The list of codeBlocks and their positions which we would like to update.
   */
  blocks: NodeWithPosition[];

  /**
   * When a delete happens within the last valid decoration in a block it causes the editor to jump. This skipLast
   * should be set to true immediately after a delete which then allows for createDecorations to skip updating the decoration
   * for the last refactor node, and hence preventing the jumpy bug.
   */
  skipLast: boolean;
}

/**
 * Retrieves positioned refractor nodes from the positionedNode
 *
 * @param nodeWithPosition - a node and position
 * @returns the positioned refractor nodes which are text, classes and a FromTo interface
 */
const getPositionedRefractorNodes = ({ node, pos }: NodeWithPosition) => {
  let startPos = pos + 1;
  const refractorNodes = refractor.highlight(node.textContent || '', node.attrs.language || 'markup');
  function mapper(refractorNode: ParsedRefractorNode): PositionedRefractorNode {
    const from = startPos;
    const to = from + refractorNode.text.length;
    startPos = to;
    return {
      ...refractorNode,
      from,
      to,
    };
  }

  const parsedRefractorNodes = parseRefractorNodes(refractorNodes);

  return flattenArray<ParsedRefractorNode>(parsedRefractorNodes).map(mapper);
};

/**
 * Creates a decoration set for the provided blocks
 */
export const createDecorations = ({ blocks, skipLast }: CreateDecorationsParams) => {
  const decorations: Decoration[] = [];

  blocks.forEach(block => {
    const positionedRefractorNodes = getPositionedRefractorNodes(block);
    const lastBlockLength = skipLast ? positionedRefractorNodes.length - 1 : positionedRefractorNodes.length;
    for (let ii = 0; ii < lastBlockLength; ii++) {
      const positionedRefractorNode = positionedRefractorNodes[ii];
      const decoration = Decoration.inline(positionedRefractorNode.from, positionedRefractorNode.to, {
        class: positionedRefractorNode.classes.join(' '),
      });
      decorations.push(decoration);
    }
  });

  return decorations;
};

interface PosWithinRangeParams extends PosParams, FromToParams {}

/**
 * Check if the position is within the range.
 */
export const posWithinRange = ({ from, to, pos }: PosWithinRangeParams) => from <= pos && to >= pos;

/**
 * Check whether the length of an array has changed
 */
export const lengthHasChanged = <GType>(prev: ArrayLike<GType>, next: ArrayLike<GType>) =>
  next.length !== prev.length;

export interface NodeInformation extends NodeTypeParams, FromToParams, ProsemirrorNodeParams, PosParams {}

/**
 * Retrieves helpful node information from the current state.
 */
export const getNodeInformationFromState = (state: EditorState): NodeInformation => {
  const { $head } = state.selection;
  const depth = $head.depth;
  const from = $head.start(depth);
  const to = $head.end(depth);
  const node = $head.parent;
  const type = node.type;
  const pos = depth > 0 ? $head.before(depth) : 0;
  return {
    from,
    to,
    type,
    node,
    pos,
  };
};

/**
 * Check that the attributes exist and are valid for the codeBlock updateAttrs.
 */
export const isValidCodeBlockAttrs = (attrs?: Attrs): attrs is CodeBlockAttrs =>
  bool(attrs && isObject(attrs) && isString(attrs.language) && attrs.language.length);

/**
 * Updates the node attrs.
 *
 * This is used to update the language for the codeBlock.
 */
export const updateNodeAttrs = (type: NodeType) => (attrs: CodeBlockAttrs): ProsemirrorCommandFunction => (
  { tr, selection },
  dispatch,
) => {
  if (!isValidCodeBlockAttrs(attrs)) {
    throw new Error('Invalid attrs passed to the updateAttrs method');
  }

  const parent = findParentNodeOfType({ types: type, selection });

  if (!parent || isEqual(attrs, parent.node.attrs)) {
    // Do nothing since the attrs are the same
    return false;
  }

  tr.setNodeMarkup(parent.pos, type, attrs);

  if (dispatch) {
    dispatch(tr);
  }

  return true;
};

const PRELOADED_LANGUAGES = [markup, clike, css, js];

/**
 * The list of strings that are recognised language names based on the the configured
 * supported languages.
 */
export const getLanguageNamesAndAliases = (supportedLanguages: RefractorSyntax[]) => {
  return uniqueArray(
    flattenArray(
      [...PRELOADED_LANGUAGES, ...supportedLanguages].map(({ name, aliases }) => [name, ...aliases]),
    ),
  );
};

/**
 * Returns true if the language is supported.
 */
export const isSupportedLanguage = (language: string, supportedLanguages: RefractorSyntax[]) => {
  return getLanguageNamesAndAliases(supportedLanguages).includes(language);
};

interface GetLanguageParams {
  /**
   * The language input from the user;
   */
  language: string;

  /**
   * The languages supported by the editor.
   */
  supportedLanguages: RefractorSyntax[];

  /**
   * The default language to use if none found.
   */
  fallback: string;
}

/**
 * Get the language from user input.
 */
export const getLanguage = ({ language, supportedLanguages, fallback }: GetLanguageParams) =>
  !isSupportedLanguage(language, supportedLanguages) ? fallback : language;

interface FormatCodeBlockFactoryParams
  extends NodeTypeParams,
    Required<Pick<CodeBlockExtensionOptions, 'formatter' | 'supportedLanguages' | 'defaultLanguage'>> {}

/**
 * A factory for creating a command which can format a selected codeBlock (or one located at the provided position).
 */
export const formatCodeBlockFactory = ({
  type,
  formatter,
  supportedLanguages,
  defaultLanguage: fallback,
}: FormatCodeBlockFactoryParams) => (
  { pos }: Partial<PosParams> = Object.create(null),
): ProsemirrorCommandFunction => (state, dispatch) => {
  const { tr, selection } = state;

  const { from, to } = pos ? { from: pos, to: pos } : selection;

  // Find the current codeBlock the cursor is positioned in.
  const codeBlock = findParentNodeOfType({ types: type, selection });

  if (!codeBlock) {
    return false;
  }

  // Get the `language`, `source` and `cursorOffset` for the block and run the formatter
  const {
    node: { attrs, textContent },
    start,
  } = codeBlock;

  const offsetStart = from - start;
  const offsetEnd = to - start;
  const language = getLanguage({ language: attrs.language, fallback, supportedLanguages });
  const formatStart = formatter({ source: textContent, language, cursorOffset: offsetStart });
  let formatEnd: FormattedContent | undefined;

  // When the user has a selection
  if (offsetStart !== offsetEnd) {
    formatEnd = formatter({ source: textContent, language, cursorOffset: offsetEnd });
  }

  if (!formatStart) {
    return false;
  }

  const { cursorOffset, formatted } = formatStart;

  // Do nothing if nothing has changed
  if (formatted === textContent) {
    return false;
  }

  const end = start + textContent.length;

  // Replace the codeBlock content with the transformed text.
  tr.insertText(formatted, start, end);

  // Set the new selection
  const anchor = start + cursorOffset;
  const head = formatEnd ? start + formatEnd.cursorOffset : undefined;

  tr.setSelection(TextSelection.create(tr.doc, anchor, head));

  if (dispatch) {
    dispatch(tr);
  }

  return true;
};

/**
 * Retrieve the supported language names based on configuration.
 */
export const getSupportedLanguagesMap = (supportedLanguages: RefractorSyntax[]) => {
  const obj: Record<string, string> = Object.create(null);
  for (const { name, aliases } of [...PRELOADED_LANGUAGES, ...supportedLanguages]) {
    obj[name] = name;
    aliases.forEach(alias => {
      obj[alias] = name;
    });
  }
  return obj;
};
