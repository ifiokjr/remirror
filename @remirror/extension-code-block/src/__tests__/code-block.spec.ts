import { fromHTML, toHTML } from '@remirror/core';
import { createBaseTestManager } from '@remirror/test-fixtures';
import { pmBuild } from 'jest-prosemirror';
import { renderEditor } from 'jest-remirror';
import typescriptPlugin from 'prettier/parser-typescript';
import { formatWithCursor } from 'prettier/standalone';
import javascript from 'refractor/lang/javascript';
import markdown from 'refractor/lang/markdown';
import tsx from 'refractor/lang/tsx';
import typescript from 'refractor/lang/typescript';

import { CodeBlockExtension, CodeBlockExtensionOptions } from '../';
import { CodeBlockFormatter } from '../code-block-types';
import { getLanguage } from '../code-block-utils';

describe('schema', () => {
  const { schema } = createBaseTestManager([{ extension: new CodeBlockExtension(), priority: 1 }]);
  const attrs = { language: 'typescript' };
  const content = 'unchanged without decorations';

  const { codeBlock, doc } = pmBuild(schema, {
    codeBlock: { nodeType: 'codeBlock', ...attrs },
  });

  it('creates the correct dom node', () => {
    expect(toHTML({ node: codeBlock(content), schema })).toBe(
      `<pre class="language-${attrs.language}"><code data-code-block-language="${attrs.language}">${content}</code></pre>`,
    );
  });

  it('parses the dom structure and finds itself', () => {
    const node = fromHTML({
      schema,
      content: `<pre><code class="language-${attrs.language}" data-code-block-language="${attrs.language}">${content}</code></pre>`,
    });
    const expected = doc(codeBlock(content));
    expect(node).toEqualProsemirrorNode(expected);
  });
});

describe('constructor', () => {
  it('is created with the correct options', () => {
    const codeBlock = new CodeBlockExtension({
      syntaxTheme: 'a11yDark',
    });
    expect(codeBlock.options.syntaxTheme).toEqual('a11yDark');
    expect(codeBlock.options.defaultLanguage).toEqual('markup');
  });
});

const supportedLanguages = [typescript, javascript, markdown, tsx];

const create = (params: CodeBlockExtensionOptions = Object.create(null)) =>
  renderEditor({
    plainNodes: [],
    attrNodes: [new CodeBlockExtension({ ...params, supportedLanguages })],
  });

describe('plugin', () => {
  let {
    view,
    add,
    attrNodes: { codeBlock },
    nodes: { doc, p },
  } = create();

  let { dom } = view;

  let tsBlock = codeBlock({ language: 'typescript' });

  beforeEach(() => {
    ({
      view,
      add,
      attrNodes: { codeBlock },
      nodes: { doc, p },
    } = create());
    tsBlock = codeBlock({ language: 'typescript' });
    ({ dom } = view);
  });

  it('renders the correct decorations', () => {
    add(doc(tsBlock(`const a = 'test';`)));

    expect(dom.querySelector('.language-typescript code')!.innerHTML).toMatchSnapshot();
  });

  it('can be updated', () => {
    const plainBlock = codeBlock({});
    const adder = add(doc(tsBlock(`const a = 'test';<cursor>`), plainBlock('Nothing to see here')));

    expect(dom.querySelector('.language-typescript code')!.innerHTML).toMatchSnapshot();
    adder.insertText('\n\nlog(a);');
    expect(dom.querySelector('.language-typescript code')!.innerHTML).toMatchSnapshot();
  });

  it('updates when multiple changes occur', () => {
    const { overwrite } = add(doc(tsBlock(`const a = 'test';`), tsBlock(`let b;`)));
    expect(dom.innerHTML).toMatchSnapshot();

    overwrite(doc(tsBlock(`const c = 'test';`), tsBlock(`let d;`)));
    expect(dom.innerHTML).toMatchSnapshot();
  });

  it('changes markup when the language changes', () => {
    const markupBlock = codeBlock();
    const content = `const a = 'test';`;
    const { overwrite } = add(doc(markupBlock(content)));
    const initialHtml = dom.querySelector('.language-markup code')!.innerHTML;

    overwrite(doc(tsBlock(content)));
    const newHtml = dom.querySelector('.language-typescript code')!.innerHTML;
    expect(newHtml).not.toBe(initialHtml);
    expect(newHtml).toMatchSnapshot();
  });

  describe('Backspace', () => {
    it('can be deleted', () => {
      const { state } = add(doc(tsBlock('<cursor>'))).press('Backspace');
      expect(state.doc).toEqualRemirrorDocument(doc(p()));
    });

    it('is still deleted when all that remains is whitespace', () => {
      const { state } = add(doc(tsBlock('<cursor>          '))).press('Backspace');
      expect(state.doc).toEqualRemirrorDocument(doc(p()));
    });

    it('steps into the previous node when content', () => {
      const { state } = add(doc(p(), tsBlock('<cursor>content')))
        .press('Backspace')
        .insertText('abc');
      expect(state.doc).toEqualRemirrorDocument(doc(p('abc'), tsBlock('content')));
    });

    it('creates a paragraph node when content', () => {
      const { state } = add(doc(tsBlock('<cursor>content')))
        .press('Backspace')
        .insertText('abc');
      expect(state.doc).toEqualRemirrorDocument(doc(p('abc'), tsBlock('content')));
    });
  });

  describe('Space', () => {
    it('responds to space input rule', () => {
      const { state } = add(doc(p('<cursor>'))).insertText('```typescript abc');
      expect(state.doc).toEqualRemirrorDocument(doc(tsBlock('abc'), p()));
    });

    it('responds to empty space input rule using the default language', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      const { state } = add(doc(p('<cursor>'))).insertText('``` abc');
      expect(state.doc).toEqualRemirrorDocument(doc(markupBlock('abc'), p()));
    });

    it('does not match invalid regex', () => {
      const { state } = add(doc(p('<cursor>'))).insertText('```123-__ ');
      expect(state.doc).toEqualRemirrorDocument(doc(p('```123-__ ')));
    });

    it('use default markup for non existent language', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      const { state } = add(doc(p('<cursor>'))).insertText('```abcde abc');
      expect(state.doc).toEqualRemirrorDocument(doc(markupBlock('abc'), p()));
    });

    it('keeps alias language name when supported', () => {
      const htmlBlock = codeBlock({ language: 'html' });
      const { state } = add(doc(p('<cursor>'))).insertText('```html abc');
      expect(state.doc).toEqualRemirrorDocument(doc(htmlBlock('abc'), p()));
    });
  });

  describe('Enter', () => {
    it('responds to enter key press', () => {
      const { state } = add(doc(p('<cursor>')))
        .insertText('```typescript')
        .press('Enter')
        .insertText('abc');
      expect(state.doc).toEqualRemirrorDocument(doc(tsBlock('abc'), p()));
    });

    it('uses default language when no language provided', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      const { state } = add(doc(p('<cursor>')))
        .insertText('```')
        .press('Enter')
        .insertText('abc');
      expect(state.doc).toEqualRemirrorDocument(doc(markupBlock('abc'), p()));
    });

    it('uses default language when given an invalid language', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      const { state } = add(doc(p('<cursor>')))
        .insertText('```invalidlang')
        .press('Enter')
        .insertText('abc');
      expect(state.doc).toEqualRemirrorDocument(doc(markupBlock('abc'), p()));
    });

    it('keeps alias language name when supported', () => {
      const htmlBlock = codeBlock({ language: 'html' });
      const { state } = add(doc(p('<cursor>')))
        .insertText('```html')
        .press('Enter')
        .insertText('abc');
      expect(state.doc).toEqualRemirrorDocument(doc(htmlBlock('abc'), p()));
    });
  });
});

describe('commands', () => {
  let {
    view,
    add,
    attrNodes: { codeBlock },
    nodes: { doc, p },
  } = create();

  let tsBlock = codeBlock({ language: 'typescript' });

  beforeEach(() => {
    ({
      view,
      add,
      attrNodes: { codeBlock },
      nodes: { doc, p },
    } = create());
    tsBlock = codeBlock({ language: 'typescript' });
  });

  describe('updateCodeBlock ', () => {
    it('updates the language', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      const { actions } = add(doc(markupBlock(`const a = 'test';<cursor>`)));

      expect(view.dom.querySelector('.language-markup code')).toBeTruthy();

      actions.updateCodeBlock({ language: 'javascript' });

      expect(view.dom.querySelector('.language-markup code')).toBeFalsy();
      expect(view.dom.querySelector('.language-javascript code')!.outerHTML).toMatchSnapshot();
    });
  });

  describe('createCodeBlock ', () => {
    it('creates the codeBlock', () => {
      const { state } = add(doc(p(`<cursor>`))).actionsCallback(actions => {
        actions.createCodeBlock({ language: 'typescript' });
      });

      expect(state.doc).toEqualRemirrorDocument(doc(tsBlock('')));
    });

    it('creates the default codeBlock when no language is provided', () => {
      const markupBlock = codeBlock({ language: 'markup' });

      const { state } = add(doc(p(`<cursor>`))).actionsCallback(actions => {
        // @ts-ignore
        actions.createCodeBlock();
      });

      expect(state.doc).toEqualRemirrorDocument(doc(markupBlock('')));
    });
  });

  describe('toggleCodeBlock ', () => {
    it('toggles the codeBlock', () => {
      const { state, actionsCallback } = add(doc(p(`<cursor>`))).actionsCallback(actions => {
        actions.toggleCodeBlock({ language: 'typescript' });
      });

      expect(state.doc).toEqualRemirrorDocument(doc(tsBlock('')));

      const { state: stateTwo } = actionsCallback(actions =>
        actions.toggleCodeBlock({ language: 'typescript' }),
      );
      expect(stateTwo.doc).toEqualRemirrorDocument(doc(p('')));
    });

    it('toggles the default codeBlock when no language is provided', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      const { state: stateOne, actionsCallback } = add(doc(p(`<cursor>`))).actionsCallback(actions => {
        actions.toggleCodeBlock();
      });

      expect(stateOne.doc).toEqualRemirrorDocument(doc(markupBlock('')));

      const { state: stateTwo } = actionsCallback(actions => actions.toggleCodeBlock());
      expect(stateTwo.doc).toEqualRemirrorDocument(doc(p('')));
    });
  });

  describe('formatCodeBlock', () => {
    const formatter: CodeBlockFormatter = ({ cursorOffset, language, source }) => {
      if (getLanguage({ fallback: 'text', language, supportedLanguages }) === 'typescript') {
        return formatWithCursor(source, {
          cursorOffset,
          plugins: [typescriptPlugin],
          parser: 'typescript',
          singleQuote: true,
        });
      }
      return;
    };

    beforeEach(() => {
      ({
        view,
        add,
        attrNodes: { codeBlock },
        nodes: { doc, p },
      } = create({ formatter }));

      tsBlock = codeBlock({ language: 'typescript' });
    });

    it('can format the codebase', () => {
      const { state } = add(
        doc(tsBlock(`const a: string\n = 'test'  ;\n\n\nlog("welcome friends")<cursor>`)),
      ).actionsCallback(actions => actions.formatCodeBlock());

      expect(state.doc).toEqualRemirrorDocument(
        doc(tsBlock(`const a: string = 'test';\n\nlog('welcome friends');\n`)),
      );
    });

    it('maintains cursor position after formatting', () => {
      const { state } = add(doc(tsBlock(`const a: string\n = 'test<cursor>'  ;\n\n\nlog("welcome friends")`)))
        .actionsCallback(actions => actions.formatCodeBlock())
        .insertText('ing');

      expect(state.doc).toEqualRemirrorDocument(
        doc(tsBlock(`const a: string = 'testing';\n\nlog('welcome friends');\n`)),
      );
    });

    it('formats text selections', () => {
      const { state, start, end } = add(
        doc(tsBlock(`<start>const a: string\n = 'test'  ;<end>\n\n\nlog("welcome friends")`)),
      ).actionsCallback(actions => actions.formatCodeBlock());

      expect(state.doc).toEqualRemirrorDocument(
        doc(tsBlock(`const a: string = 'test';\n\nlog('welcome friends');\n`)),
      );
      expect([start, end]).toEqual([1, 26]);
    });

    it('can format complex scenarios', () => {
      const content = p('Hello darkness, my old friend.');
      const otherCode = tsBlock(`document.addEventListener("click",  log)`);
      const { state } = add(
        doc(
          content,
          content,
          tsBlock(`const a: string\n = 'test<cursor>'  ;\n\n\nlog("welcome friends")`),
          content,
          content,
          otherCode,
        ),
      )
        .actionsCallback(actions => actions.formatCodeBlock())
        .insertText('ing');

      expect(state.doc).toEqualRemirrorDocument(
        doc(
          content,
          content,
          tsBlock(`const a: string = 'testing';\n\nlog('welcome friends');\n`),
          content,
          content,
          otherCode,
        ),
      );
    });
  });
});
