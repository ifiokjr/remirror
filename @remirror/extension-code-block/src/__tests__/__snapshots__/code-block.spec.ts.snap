// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin can be updated 1`] = `"<span class=\\"token keyword\\">const</span><span> a </span><span class=\\"token operator\\">=</span><span> </span><span class=\\"token string\\">'test'</span><span class=\\"token punctuation\\">;</span>"`;

exports[`plugin can be updated 2`] = `
"<span class=\\"token keyword\\">const</span><span> a </span><span class=\\"token operator\\">=</span><span> </span><span class=\\"token string\\">'test'</span><span class=\\"token punctuation\\">;</span><span>

</span><span class=\\"token builtin\\">console</span><span class=\\"token punctuation\\">.</span><span class=\\"token function\\">log</span><span class=\\"token punctuation\\">(</span><span>a</span><span class=\\"token punctuation\\">)</span>;"
`;

exports[`plugin renders the correct decorations 1`] = `"<span class=\\"token keyword\\">const</span><span> a </span><span class=\\"token operator\\">=</span><span> </span><span class=\\"token string\\">'test'</span><span class=\\"token punctuation\\">;</span>"`;

exports[`plugin updates when multiple changes occur 1`] = `"<pre><code class=\\"language-typescript\\" data-code-block-language=\\"typescript\\"><span class=\\"token keyword\\">const</span><span> a </span><span class=\\"token operator\\">=</span><span> </span><span class=\\"token string\\">'test'</span><span class=\\"token punctuation\\">;</span></code></pre><pre><code class=\\"language-typescript\\" data-code-block-language=\\"typescript\\"><span class=\\"token keyword\\">let</span><span> b</span><span class=\\"token punctuation\\">;</span></code></pre>"`;

exports[`plugin updates when multiple changes occur 2`] = `"<pre><code class=\\"language-typescript\\" data-code-block-language=\\"typescript\\"><span class=\\"token keyword\\">const</span><span> c </span><span class=\\"token operator\\">=</span><span> </span><span class=\\"token string\\">'test'</span><span class=\\"token punctuation\\">;</span></code></pre><pre><code class=\\"language-typescript\\" data-code-block-language=\\"typescript\\"><span class=\\"token keyword\\">let</span><span> d</span><span class=\\"token punctuation\\">;</span></code></pre>"`;
